### ISSUE TITLE
Expose array data to python through buffer protocol

### ISSUE THREAD

**Issue by kylebarron** [developmentseed/async-tiff#146]
it should be possible to set the buffer protocol metadata correctly so that we can store just a Vec<u8> on the rust side that automatically gets exposed as a numpy array with the right dimension and dtype on the python side

**Comment by weiji14** [comment]
> automatically gets exposed as a numpy array with the right dimension and dtype on the python side

How would this be done? The [`np.frombuffer`](https://numpy.org/doc/2.4/reference/generated/numpy.frombuffer.html) docs seem to indicate that buffers will be converted to a 1-D array only. The dtype also needs to be passed in as an argument.

Maybe we should go through [DLPack](https://dmlc.github.io/dlpack/latest/python_spec.html) instead which includes the shape, dtype and other attributes within the struct, and is supported more widely:
- [`np.from_dlpack`](https://numpy.org/doc/2.4/reference/generated/numpy.from_dlpack.html)
- [`cupy.from_dlpack`](https://docs.cupy.dev/en/v14.0.0rc1/reference/generated/cupy.from_dlpack.html)
- [`torch.utils.dlpack.from_dlpack`](https://docs.pytorch.org/docs/2.9/dlpack.html#torch.utils.dlpack.from_dlpack)
- etc

We could either use [`dlpark`](https://crates.io/crates/dlpark) to convert Vec -> ndarray -> DLPack, or an alternative implementation at [`dlpk`](https://crates.io/crates/dlpk). Zarrs uses `dlpark` (see https://github.com/zarrs/zarrs/issues/113) and I'm more comfortable with it, though technically the [DLPack structs](https://dmlc.github.io/dlpack/latest/c_api.html#structs) should be cross-compatible if implemented correctly.

Another thought is to **not** implement this, or treat it as low priority, given Jeff's warnings at https://github.com/developmentseed/async-tiff/discussions/7 about how it was a mistake in aiocogeo to have higher-level functions to return arrays 🙂

**Comment by kylebarron** [comment]
#164 is an example of exposing a 2d buffer through the buffer protocol. This is hard-coded, but it works for instructional purposes.

```py
from async_tiff._async_tiff import example_array
import numpy as np

# Create array in Rust memory
rust_array = example_array()

# create numpy view of the Rust memory using buffer protocol
np_view = np.array(rust_array)

# Print numpy array
np_view
# array([[1, 2, 3],
#        [4, 5, 6]], dtype=uint8)
```

In a real implementation we'll ensure that the data type, stride, dimension, shape, etc is correct as specified in the underlying array.

DLPack is _much, much_ more complex than we need. I would be quite reluctant to add DLPack integration.

**Comment by kylebarron** [comment]
> Another thought is to **not** implement this, or treat it as low priority, given Jeff's warnings at [#7](https://github.com/developmentseed/async-tiff/discussions/7) about how it was a mistake in aiocogeo to have higher-level functions to return arrays 🙂

It's a good opportunity to look back at #7 for a read, after this point of implementation.

I think the goal of this crate should be to expose a useful, simple API for accessing the raw, uncompressed data of individual tiles from a TIFF. That means that this crate should have _no interpretation_ of what any of the bytes mean.

This means we _cannot_ expose a masked array, because that would require us to interpret what the TIFF tags mean in the context of nodata. But we should expose the tags that would allow a higher-level library (`async-geotiff`?) to wrap `async-tiff` to apply those interpretations.

### Result should be decompressed

_Decompressed_ data means that the primary API **should** expose the result of each tile after it's been decompressed and the predictor reversed.

(Though we should and I think do expose an alternate API for tile content before it's been decompressed)

### Result should be as easy to use as possible while not adding abstractions

This means that we **should** expose the tiff tiles to the Python user as multi-dimensional arrays with the correct data type.

Every TIFF IFD has a non-nullable tag called [`SamplesPerPixel`](https://github.com/developmentseed/async-tiff/blob/dd8f2d0f9068846c941cc331132db3ddd74df372/src/ifd.rs#L54) that defines the "depth" of the given array. For example, a grayscale image would have `SamplesPerPixel: 1` while an RGB image would have `SamplesPerPixel: 3`.

- Question: Is a tiff array with one sample exposed as 2D or do we expose all tiles as 3D? I think it would probably be better to always expose the array as 3D for consistency.

#### Band-interleaved vs pixel-interleaved

We'll need to document that users must interpret the array based on the value of the `PlanarConfiguration` tag:

- Pixel-interleaved: array interpretation is `(height, width, bands)`
- Band-interleaved: array interpretation is `(bands, height, width)`

Exposing the data in this way ensures that we always expose the data as zero-copy.

In either way, users may have to reshape the data to be [in line with rasterio](https://rasterio.readthedocs.io/en/stable/topics/image_processing.html#image-processing-software). 

### Multi-dimensional buffer protocol should be primary, but not only API

Users should be able to call `.buffer` on the decoded result to get a reference to the underlying `Vec<u8>` on the rust side, which would then be interpreted by numpy as a 1D array.

cc @geospatial-jeff in case he has any thoughts.

**Comment by weiji14** [comment]
> DLPack is _much, much_ more complex than we need. I would be quite reluctant to add DLPack integration.

I'll just add one more point that it enables not just Rust<->Python interop (which the Python Buffer Protocol is limited to), but also (theoretically) other languages/frameworks. See e.g. [DLPack.jl](https://github.com/p-zubieta/DLPack.jl) and [this thread on WebGPU](https://github.com/dmlc/dlpack/issues/124).

But yeah, it is complex, so ok with going through the buffer protocol. One more question though, what's the pros/cons of your implementation in #164 vs going through ndarray (#151) and the [`numpy`](https://github.com/PyO3/rust-numpy) crate?

**Comment by kylebarron** [comment]
> I'll just add one more point that it enables not just Rust<->Python interop

Good to know; but for the foreseeable future I think this project will only bind to Python. 

> One more question though, what's the pros/cons of your implementation in #164 vs going through ndarray (#151) and the [`numpy`](https://github.com/PyO3/rust-numpy) crate?

The buffer protocol is more general than numpy-specific integration. We still may consider an ndarray integration for rust users, but I think exposing as buffer protocol is preferred in Python.
